File: proxy-go/handler/circuit.go
package handler

import (
	"os"
	"strconv"
	"sync"
	"time"
)

type cbState int

const (
	cbClosed cbState = iota // normal operation
	cbOpen                  // failing — reject all requests
	cbHalf                  // trial — let one through
)

type circuitBreaker struct {
	mu         sync.Mutex
	state      cbState
	failures   int
	lastFailed time.Time
	threshold  int
	timeout    time.Duration
}

var (
	cbMu      sync.RWMutex
	breakers  = make(map[string]*circuitBreaker)
	cbThresh  int
	cbTimeout time.Duration
)

func init() {
	cbThresh = 5
	if v := os.Getenv("TSM_CB_THRESHOLD"); v != "" {
		if n, _ := strconv.Atoi(v); n > 0 {
			cbThresh = n
		}
	}
	cbTimeout = 30 * time.Second
	if v := os.Getenv("TSM_CB_TIMEOUT_MS"); v != "" {
		if n, _ := strconv.Atoi(v); n > 0 {
			cbTimeout = time.Duration(n) * time.Millisecond
		}
	}
}

func getBreaker(upstream string) *circuitBreaker {
	cbMu.RLock()
	cb := breakers[upstream]
	cbMu.RUnlock()
	if cb != nil {
		return cb
	}
	cbMu.Lock()
	defer cbMu.Unlock()
	if cb = breakers[upstream]; cb != nil {
		return cb
	}
	cb = &circuitBreaker{threshold: cbThresh, timeout: cbTimeout}
	breakers[upstream] = cb
	return cb
}

// isAllowed returns true if the circuit is CLOSED or entering HALF state.
func isAllowed(upstream string) bool {
	cb := getBreaker(upstream)
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case cbClosed:
		return true
	case cbOpen:
		if time.Since(cb.lastFailed) > cb.timeout {
			cb.state = cbHalf
			return true
		}
		return false
	case cbHalf:
		return true
	}
	return true
}

func recordSuccess(upstream string) {
	cb := getBreaker(upstream)
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.failures = 0
	cb.state = cbClosed
}

func recordFailure(upstream string) {
	cb := getBreaker(upstream)
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.failures++
	cb.lastFailed = time.Now()
	if cb.failures >= cb.threshold {
		cb.state = cbOpen
	}
}

// breakerStatus returns a snapshot of all circuit breakers for the health endpoint.
func breakerStatus() map[string]string {
	cbMu.RLock()
	defer cbMu.RUnlock()
	out := make(map[string]string, len(breakers))
	for name, cb := range breakers {
		cb.mu.Lock()
		switch cb.state {
		case cbClosed:
			out[name] = "closed"
		case cbOpen:
			out[name] = "open"
		case cbHalf:
			out[name] = "half"
		}
		cb.mu.Unlock()
	}
	return out
}

File: proxy/src/circuit.ts
/**
 * Per-upstream circuit breaker.
 *
 * States:
 *   CLOSED  — normal operation, requests flow through
 *   OPEN    — upstream is failing, requests fail fast
 *   HALF    — probe request allowed; if it succeeds → CLOSED
 *
 * Config (env):
 *   TSM_CB_THRESHOLD   — consecutive failures before OPEN (default 5)
 *   TSM_CB_TIMEOUT_MS  — ms to wait in OPEN before probing (default 30000)
 */

type State = 'CLOSED' | 'OPEN' | 'HALF';

interface Breaker {
  state:      State;
  failures:   number;
  openedAt:   number;
}

const THRESHOLD  = parseInt(process.env.TSM_CB_THRESHOLD  ?? '5');
const TIMEOUT_MS = parseInt(process.env.TSM_CB_TIMEOUT_MS ?? '30000');

const breakers = new Map<string, Breaker>();

function get(upstream: string): Breaker {
  if (!breakers.has(upstream)) {
    breakers.set(upstream, { state: 'CLOSED', failures: 0, openedAt: 0 });
  }
  return breakers.get(upstream)!;
}

/** Returns true if a request is allowed through for this upstream. */
export function isAllowed(upstream: string): boolean {
  const b = get(upstream);
  if (b.state === 'CLOSED') return true;
  if (b.state === 'OPEN') {
    if (Date.now() - b.openedAt >= TIMEOUT_MS) {
      b.state = 'HALF';
      return true;   // probe
    }
    return false;    // fail fast
  }
  return true;       // HALF: allow probe
}

/** Record a successful upstream response. */
export function recordSuccess(upstream: string): void {
  const b = get(upstream);
  b.failures = 0;
  b.state    = 'CLOSED';
}

/** Record an upstream failure. Opens the breaker if threshold exceeded. */
export function recordFailure(upstream: string): void {
  const b = get(upstream);
  b.failures++;
  if (b.state === 'HALF' || b.failures >= THRESHOLD) {
    b.state    = 'OPEN';
    b.openedAt = Date.now();
  }
}

/** Breaker status for the /health and /metrics endpoints. */
export function breakerStatus(): Record<string, { state: State; failures: number }> {
  const out: Record<string, { state: State; failures: number }> = {};
  for (const [k, v] of breakers) {
    out[k] = { state: v.state, failures: v.failures };
  }
  return out;
}
